text/0000-impl-trait-type-alias.md

```diff
@@ -0,0 +1,624 @@
+      // Creates a variable `x` of type `i32`, which is equal to type `Foo`
+      let x: i32 = foo();
+      x + 5
+  }
+}
+
+fn use_foo_outside_mod() {
+    // Creates a variable `x` of type `Foo`, which is equal to type `impl Debug`
+    let x = my_mod::foo();
+
+    // Because we're outside `my_mod`, using a value of type `Foo` as anything
+    // other than `impl Debug` is an error:
+    let y: i32 = my_mod::foo(); // ERROR: expected type `i32`, found type `Foo`
+}
+```
+
+This makes it possible to write modules that hide their concrete types from the
+outside world, allowing them to change implementation details without affecting
+consumers of their API.
+
+Note that it is sometimes necessary to manually specify the concrete type of an
+`impl Trait`-aliased value, like in `let x: i32 = foo();` above.
+This aids the function's ability to locally infer the concrete type of `Foo`.
+
+One particularly noteworthy use of `impl Trait` type aliases is in trait
+implementations.
+With this feature, we can declare `impl Trait` associated types:
+
+```rust
+struct MyType;
+impl Iterator for MyType {
+    type Item = impl Debug;
+    fn next(&mut self) -> Option<Self::Item> {
+        Some("Another item!")
+    }
+}
+```
+
+In this trait implementation, we've declared that the item returned by our
+iterator implements `Debug`, but we've kept its concrete type (`&'static str`)
+hidden from the outside world.
+
+We can even use this feature to specify unnameable associated types, such as
+closures:
+
+```rust
+struct MyType;
+impl Iterator for MyType {
+    type Item = impl Fn(i32) -> i32;
+    fn next(&mut self) -> Option<Self::Item> {
+        Some(|x| x + 5)
+    }
+}
+```
+
+`impl Trait` aliases can also be used to reference unnameable types in a struct
+definition:
+
+```rust
+type Foo = impl Debug;
+fn foo() -> Foo { 5i32 }
+
+struct ContainsFoo {
+    some_foo: Foo
+}
+```
+
+
+`impl Trait` can also appear inside of another type in a type alias:
+
+```rust
+type Foo = Option<impl Debug>;
+fn foo() -> Foo {
+    Some("Debuggable")
+}
+```
+
+Or even multiple times within the same type alias:
+
+```rust
+type Foo = (impl Debug, impl Fn());
+fn foo() -> Foo {
+    ("Debuggable", || println!("Hello, world!"))
+}
+```
+
+It's also possible to write generic `impl Trait` aliases:
+
+```rust
+#[derive(Debug)]
+struct MyStruct<T: Debug> {
+    inner: T
+};
+
+type Foo<T> = impl Debug;
+
+fn get_foo<T: Debug>(x: T) -> Foo<T> {
+    MyStruct {
+        inner: x
+    }
+}
+```
+
+As specified in
+[RFC 1951](https://github.com/rust-lang/rfcs/blob/master/text/1951-expand-impl-trait.md),
+`impl Trait` implicitly captures all generic types parameters in scope.
+In practice, this means that `impl Trait` associated types may contain generic
+types from an impl:
+
+```rust
+struct MyStruct;
+trait Foo<T> {
+    type Bar;
+    fn bar() -> Bar;
+}
+
+impl<T> Foo<T> for MyStruct {
+    type Bar = impl Trait;
+    fn bar() -> impl Trait {
```

Comment:
hmm, is it a typo that you wrote `fn bar -> impl Trait` here instead of `fn bar() -> Bar` or `fn bar() -> Self::Bar` ?

In particular, I would have expected that a separate method that has the return type `impl Trait` might have a totally different concrete implementation type than that uses for the associated `type Bar = impl Trait`... and thus the signature `fn bar() -> impl Trait` does not match the already given `fn bar() -> Bar`.

----

(of course there is no *precedent* previously set here, since under [conservative impl trait](https://github.com/rust-lang/rfcs/blob/master/text/1522-conservative-impl-trait.md) we do not support returning `impl Trait` outside of free functions and inherent methods. But in terms of what my *intuition* would lead me to think, I would have guessed that it is not legal to write a trait method with the signature `fn bar() -> impl Trait` even with this new RFC in place.)

Reply:
Yep, this is absolutely a typo.

Reply:
Hmm, after reading the entirety of the comment thread, I am no longer sure that this was a mere typo.

Reply:
oh okay, now that I see @cramertj 's note, I guess it was a typo. 😜

Reply:
Yeah, "mere typo" is probably a bad characterization. I hadn't thought as deeply as I should have about the interactions involved in this situation. It's something I'd still like to make work down the road. It could also be addressed through something like associated-type inference.